gui/src/shortcutdialog.cpp

```diff
@@ -0,0 +1,311 @@
+        artwork["hero"] = artworkWidgets.value(ArtworkType::HERO)->getUrl().toStdString();
+        artwork["logo"] = artworkWidgets.value(ArtworkType::LOGO)->getUrl().toStdString();
+        artwork["icon"] = artworkWidgets.value(ArtworkType::ICON)->getUrl().toStdString();
+        CreateShortcut(server, artwork);
+    });
+
+    //QTimer::singleShot(0, this, &ShortcutDialog::dialogLoaded);
+}
+
+void ShortcutDialog::ExternalChanged() {
+    Ui::ShortcutDialog::external_dns_edit->setFocus();
+    Ui::ShortcutDialog::external_dns_edit->setEnabled(Ui::ShortcutDialog::allow_external_access_checkbox->isChecked());
+    Ui::ShortcutDialog::local_ssid_edit->setEnabled(Ui::ShortcutDialog::allow_external_access_checkbox->isChecked());
+}
+
+std::string ShortcutDialog::getExecutable() {
+#if defined(__APPLE__)
+    char buffer[PATH_MAX];
+    uint32_t size = sizeof(buffer);
+
+    if (_NSGetExecutablePath(buffer, &size) == -1) {
+        // Buffer is not large enough, allocate a larger one
+        char* largerBuffer = new char[size];
+        if (_NSGetExecutablePath(largerBuffer, &size) == 0) {
+            return largerBuffer;
+        }
+    } else {
+        return buffer;
+    }
+#elif defined(__linux)
+    std::string chiakiFlatpakDir = "";
+
+    chiakiFlatpakDir.append(getenv("HOME"));
+    chiakiFlatpakDir.append("/.var/app/io.github.streetpea.Chiaki4deck");
+    //Installed via Flatpak
+    if (access(chiakiFlatpakDir.c_str(), 0) == 0) {
+        return "flatpak run io.github.streetpea.Chiaki4deck";
+    }
+    else {
+        char buffer[PATH_MAX];
+        ssize_t len = readlink("/proc/self/exe", buffer, sizeof(buffer) - 1);
+
+        if (len != -1) {
+            buffer[len] = '\0';
+            return buffer;
+        }
+    }
+#endif
+    return "";
+}
+
+void ShortcutDialog::CreateShortcut(const DisplayServer* displayServer, std::map<std::string, std::string> artwork) {
+    //Create a map of variables for the template
+    std::map<std::string, std::string> paramMap;
+
+    paramMap["home_ssid"] = local_ssid_edit->text().toStdString();
+    paramMap["ps_console"] = displayServer->IsPS5() ? "5" : "4";
+    paramMap["wait_timeout"] = "35";
+    paramMap["home_addr"] = displayServer->GetHostAddr().toStdString();
+    paramMap["away_addr"] = external_dns_edit->text().toStdString();
+    paramMap["regist_key"] = QString::fromUtf8(displayServer->registered_host.GetRPRegistKey()).toStdString();
+    paramMap["login_passcode"] = passcode_edit->text().toStdString();
+    paramMap["mode"] = mode_combo_box->currentText().toStdString();
+    paramMap["server_nickname"] = displayServer->discovered ? displayServer->discovery_host.host_name.toStdString() : displayServer->registered_host.GetServerNickname().toStdString();
+    paramMap["executable"] = getExecutable();
+    std::stringstream escapeCommandStream;
+    escapeCommandStream << "printf %q " << "\"" << paramMap["server_nickname"] << "\"";
```

Comment:
Why would you ever need shell for this?

Reply:
@streetpea this was in your PR, was there a reason for doing this in shell rather than native C?